Stop folder when running out of disk space (fixes #2057)

Zillode opened the pull request.

calmh commented:
To add the new dependency:

```
go run build.go setup
go run build.go deps
```

calmh commented on internal/config/config.go:
```diff
@@ -462,6 +466,14 @@ func ChangeRequiresRestart(from, to Configuration) bool {
+func convertV10V11(cfg *Configuration) {
+	// Set minimum disk space on existing folders to 500MB
+	for i := range cfg.Folders {
+		cfg.Folders[i].MinimumDiskSpaceMB = 500
```
Not sure I like going by MB and defaulting to this. What about when if we sync 10 MB of files into an empty 256 MB flash card on an embedded system - shouldn't that work out of the box?

calmh commented on internal/osutil/osutil.go:
```diff
@@ -210,3 +213,8 @@ func init() {
 	return execExts[strings.ToLower(filepath.Ext(path))]
 }
+
+func FreeDiskSpaceMB(path string) (uint64, error) {
```
I suggest `int64` and just leaving it as bytes here. (We don't use unsigned ints for arithmetic, for safety reasons.)

calmh commented on internal/model/model.go:
```diff
@@ -1706,6 +1706,8 @@ func (m *Model) CheckFolderHealth(id string) error {
 			err = errors.New("folder path missing")
 		} else if !folder.HasMarker() {
 			err = errors.New("folder marker missing")
+		} else if space, syserr := osutil.FreeDiskSpaceMB(folder.Path()); syserr != nil && space < uint64(folder.MinimumDiskSpaceMB) {
```
We might still fill the disk if we're trying to pull a file that's larger than 500 MB.

calmh commented on internal/config/config.go:
```diff
@@ -67,19 +67,20 @@ func (cfg Configuration) Copy() Configuration {
+	RescanIntervalS  int                         `xml:"rescanIntervalS,attr" json:"rescanIntervalS"`
+	IgnorePerms      bool                        `xml:"ignorePerms,attr" json:"ignorePerms"`
+	AutoNormalize    bool                        `xml:"autoNormalize,attr" json:"autoNormalize"`
+	MinimumDiskFreeP int                         `xml:"MinimumDiskFreeP,attr" json:"MinimumDiskFreeP"`
```
I think `MinDiskFreePct` and `minDiskFreePct` as serialization

calmh commented on internal/config/wrapper.go:
```diff
@@ -96,6 +96,10 @@ func Load(path string, myID protocol.DeviceID) (*Wrapper, error) {
 	return Wrap(path, cfg), nil
 }
 
+func (w *Wrapper) Path() string {
```
This is the path of the config file? I think you want the `FolderConfiguration.Path()`?

calmh commented on internal/model/model.go:
```diff
@@ -1653,6 +1653,14 @@ func (m *Model) BringToFront(folder, file string) {
 func (m *Model) CheckFolderHealth(id string) error {
+	// This could be expensive wrt sleeping disks... if random(1) < 0.2 ?
+	usage, syserr := osutil.DiskUsage(m.cfg.Path())
+	if syserr == nil && usage > 99 {
```
Should use configured value here I think. :)

calmh commented on internal/osutil/osutil.go:
```diff
@@ -210,3 +211,13 @@ func init() {
 	return execExts[strings.ToLower(filepath.Ext(path))]
 }
+
+func DiskDetails(path string) (int64, int64, error) {
```
I'm not super excited about this name as it's not clear what it returns exactly. It could be clarified by something like

`func DiskUsage(path string) (size, free int64, err error)`

and then have the other one below be `func DiskUsedPercent(path string) float64` or so.

calmh commented on internal/model/rwfolder.go:
```diff
@@ -492,6 +494,18 @@ func (p *rwFolder) pullerIteration(ignores *ignore.Matcher) int {
+		folder, ok := p.model.cfg.Folders()[p.folder]
+		if ok {
+			diskSize, diskFree, diskErr := osutil.DiskDetails(p.folder)
+			if diskErr == nil && (diskSize-diskFree)-pullFileSize < (diskSize/100)*int64(folder.MinimumDiskFreeP) {
```
Should this simply be `if diskFree < pullFileSize`?

Zillode replied:
Could work, but we could anticipate it sooner with MinDiskFreePct. Your call

Zillode commented:
Do not merge, wait for upstream patch: https://github.com/ricochet2200/go-disk-usage/pull/1

Zillode commented:
@st-jenkins run integrations tests on windows please

calmh commented on internal/model/rwfolder.go:
```diff
@@ -492,6 +494,18 @@ func (p *rwFolder) pullerIteration(ignores *ignore.Matcher) int {
+		folder, ok := p.model.cfg.Folders()[p.folder]
+		if ok {
+			diskSize, diskFree, diskErr := osutil.DiskUsage(p.folder)
+			if diskErr == nil && (diskSize-diskFree)-pullFileSize < (diskSize/100)*int64(folder.MinDiskFreePct) {
```
This is too much arithmetic in the if clause, I find it hard to follow and I'm not sure it's correct. The current amount of used space (size - free), minus the space we need, less than the percentage specified?

I would expect the checkFolderHealth thing to take care of the percentage, and here just check if we have space for the files we are about to pull, i.e. something like diskFree < pullFileSize?

calmh commented on internal/model/model.go:
```diff
@@ -1653,6 +1653,14 @@ func (m *Model) BringToFront(folder, file string) {
 func (m *Model) CheckFolderHealth(id string) error {
+	// This could be expensive wrt sleeping disks... if random(1) < 0.2 ?
+	usage, syserr := osutil.DiskUsedPercentage(m.cfg.ConfigPath())
+	if syserr == nil && usage > 99 {
```
syserr is an unusual variable name, would `if usage, err := osutil.DiskUsedPercentage(...); err == nil && usage > theConfigValue {` suffice here?

Zillode replied:
1. I think we should panic here, the db could become corrupted at any time now. It's just that the panic gets recovered somehow. But if you don't like the panic I won't investigate it
2. What about wrapping it with ```if rand.Float64() < 0.2``` to keep things cheap? Should I add it?

calmh commented on internal/model/model.go:
```diff
@@ -1668,6 +1676,8 @@ func (m *Model) CheckFolderHealth(id string) error {
 			err = errors.New("folder path missing")
 		} else if !folder.HasMarker() {
 			err = errors.New("folder marker missing")
+		} else if usage, syserr := osutil.DiskUsedPercentage(folder.Path()); syserr == nil && usage > float64(100-folder.MinDiskFreePct) {
```
This is fine, apart from s/syserr/err/ I think

Zillode commented:
Strange, netbsd fails to build? Works on my machine non-dockerized.

Zillode commented:
@st-jenkins run integration tests on linux and windows please

Zillode commented:
I think we need a beta channel to properly test this before deploying it to everyone:)
Maybe we can use the strategy of android, namely delivering the new release to the lucky 5%?

calmh commented:
It needs to actually build though. :) I don't particularly care if the functionality is just stubbed out for netbsd to start with, if it works for the major platforms.

Zillode commented:
Would it be better if ```calmh/du``` provides an error if the platform is not supported?
Without this we need to list the supported platforms here, which could become out of sync with the upstream package.

calmh commented:
I had totally forgotten that I had written that package. Yes, I guess it should!

calmh commented:
@Zillode calmh/du now builds on everything, but always returns an error on netbsd, openbsd and solaris.

Zillode commented on internal/model/model.go:
```diff
@@ -1658,13 +1662,19 @@ func (m *Model) BringToFront(folder, file string) {
 	}
 
 	fi, err := os.Stat(folder.Path())
-	if v, ok := m.CurrentLocalVersion(id); ok && v > 0 {
```
Why is this `v > 0` ?

AudriusButkevicius replied:
When you initially add a folder, it doesn't have any files, hence version is 0.

Zillode commented:
Rebased and updated deps, be sure to properly test this after merging please. It works on my machine.
See the following commands:
```
mount -t tmpfs -o size=100m tmpfs /mnt/ramdisk
# Add a folder with the /mnt/ramdisk path
cd /mnt/ramdisk
dd if=/dev/zero of=file.txt count=1022222 bs=1024
```


Should we announce this to events.Default as well?

calmh commented:
Two more things from mobile (so no line comments for some reason) :) I don't think we handle the `default` annotation in FolderConfiguration so the default needs to be set (and be editable?) in the GUI. And I think DiskUsedPercentage is easier defined as 100-DiskFreePercentage

calmh commented:
I suggest https://github.com/calmh/syncthing/commit/29646e7e600207d6b837ce89434093925686185e, plus the GUI/config stuff

Zillode commented:
@st-jenkins run integration tests on linux and windows please

calmh commented:
rebased & merged
